File: python/13_tensor.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Tensor definition examples
"""

import numpy as np


def main():
    print("vector (1-leg tensor)")
    t1 = np.array([1, 2, 3])
    print(f"t1 = {t1}")
    print(f"t1.shape = {t1.shape}")
    print(f"t1[0] = {t1[0]}")
    print(f"t1[1] = {t1[1]}\n")

    print("(random) matrix (2-leg tensor)")
    t2 = np.random.rand(2, 3)
    print(f"t2 =\n{t2}")
    print(f"t2.shape = {t2.shape}")
    print(f"t2[0,0] = {t2[0, 0]}")
    print(f"t2[1,2] = {t2[1, 2]}\n")

    print("3D array (3-leg tensor)")
    A = np.array([[[1, 2], [3, 4]], [[5, 6], [7, 8]]])
    print(f"A =\n{A}\n")
    print(f"A.shape = {A.shape}")
    print(f"A[0,0,0] = {A[0, 0, 0]}")
    print(f"A[1,1,0] = {A[1, 1, 0]}")
    print(f"A[:,0,1] = {A[:, 0, 1]}")


if __name__ == "__main__":
    main()

File: julia/13_tensor.jl
#!/usr/bin/env julia
"""
Tensor definition examples
"""

using Random

function main()
    println("vector (1-leg tensor)")
    t1 = [1, 2, 3]
    println("t1 = $t1")
    println("t1.shape = ", size(t1))
    println("t1[1] = ", t1[1])   # Julia は 1-based index
    println("t1[2] = ", t1[2], "\n")

    println("(random) matrix (2-leg tensor)")
    t2 = rand(2, 3)
    println("t2 =\n", t2)
    println("t2.shape = ", size(t2))
    println("t2[1,1] = ", t2[1, 1])
    println("t2[2,3] = ", t2[2, 3], "\n")

    println("3D array (3-leg tensor)")
    A = reshape([1, 2, 3, 4, 5, 6, 7, 8], 2, 2, 2)
    A = permutedims(A, (3, 2, 1))
    println("A =\n", A, "\n")
    println("A.shape = ", size(A))
    println("A[1,1,1] = ", A[1, 1, 1])
    println("A[2,2,1] = ", A[2, 2, 1])
    println("A[:,1,2] = ", A[:, 1, 2])
end

main()
